### PR TITLE
EMSI: clean up usage of older term and support better role assignment creation 

### PR DESCRIPTION
**Pull request by yugangw-msft**
2 parts:
1. Do not use `explicit identity`, the right term should be `user assigned identity`
2. When create a role assignment for it, expose `--assignee-object-id` to bypass the graph query. This is needed when login as a service principal which by default has no graph permissions.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by yugangw-msft** [comment]
//cc: @skwan

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/tests/role_scenario_test.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+class RoleScenarioTest(ScenarioTest):
+
+    def enable_large_payload(self, size=8192):
+        from azure_devtools.scenario_tests import LargeResponseBodyProcessor
+        large_resp_body = next((r for r in self.recording_processors if isinstance(
+            r, LargeResponseBodyProcessor)), None)
+        if large_resp_body:
+            large_resp_body._max_response_body = size   # pylint: disable=protected-access
+
+    def get_guid_gen_patch(self, guids, test_seam='azure.cli.command_modules.role.custom._gen_guid'):
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/tests/role_scenario_test.py]: This is to ensure we will have fixed guid (used in the url) recorded into the .yaml files for playback later

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -34,6 +37,8 @@
 
 _CUSTOM_RULE = 'CustomRole'
 
+# pylint: disable=too-many-lines
```
Comment (troydai) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: nit: this is the indication that this file should be split into smaller ones.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 3]
```diff
@@ -118,8 +123,12 @@ def _search_role_definitions(definitions_client, name, scope, custom_role_only=F
     return roles
 
 
-def create_role_assignment(cmd, role, assignee, resource_group_name=None, scope=None):
-    return _create_role_assignment(cmd.cli_ctx, role, assignee, resource_group_name, scope)
+def create_role_assignment(cmd, role, assignee=None, assignee_object_id=None, resource_group_name=None, scope=None):
+    if bool(assignee) == bool(assignee_object_id):
+        raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
+    resolve_assignee = not assignee_object_id
```
Comment (troydai) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: nit: Can this assignment just be rolled into the parameter assignment on line 131?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 4]
```diff
@@ -184,13 +193,16 @@ def list_role_assignments(cmd, assignee=None, role=None, resource_group_name=Non
...
-        for i in [r for r in results if not r['properties'].get('principalName')]:
-            i['properties']['principalName'] = ''
-            if principal_dics.get(i['properties']['principalId']):
-                i['properties']['principalName'] = principal_dics[i['properties']['principalId']]
+            for i in [r for r in results if not r['properties'].get('principalName')]:
+                i['properties']['principalName'] = ''
+                if principal_dics.get(i['properties']['principalId']):
+                    i['properties']['principalName'] = principal_dics[i['properties']['principalId']]
+        except (CloudError, GraphErrorException):
+            pass   # failure on resolving principal due to graph permission should not block the whole thing
```
Comment (troydai) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: worth log a warning?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 2 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: From my experience working with modules that did split their custom commands into multiple files, it simply added complexity.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/_params.py [hunk 5]
```diff
@@ -100,6 +100,7 @@ def load_arguments(self, _):
         c.argument('show_all', options_list=['--all'], action='store_true', help='show all assignments under the current subscription')
         c.argument('include_inherited', action='store_true', help='include assignments applied on parent scopes')
         c.argument('assignee', help='represent a user, group, or service principal. supported format: object id, user sign-in name, or service principal name')
+        c.argument('assignee_object_id', help="assignee's graph object id, such as the 'principal id' from a managed service identity. Use this instead of '--assignee' to by-pass graph permission issues")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/_params.py]: nit: by-pass should be "bypass".

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 6]
```diff
@@ -118,8 +123,12 @@ def _search_role_definitions(definitions_client, name, scope, custom_role_only=F
     return roles
 
 
-def create_role_assignment(cmd, role, assignee, resource_group_name=None, scope=None):
-    return _create_role_assignment(cmd.cli_ctx, role, assignee, resource_group_name, scope)
+def create_role_assignment(cmd, role, assignee=None, assignee_object_id=None, resource_group_name=None, scope=None):
+    if bool(assignee) == bool(assignee_object_id):
+        raise CLIError('usage error: --assignee STRING | --assignee-object-id GUID')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Is there a reason we couldn't simply accept a string or GUID for assignee? This seems like an Xplat anti-pattern.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/tests/role_scenario_test.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+# --------------------------------------------------------------------------------------------
+import uuid
+import mock
+from azure.cli.testsdk import ScenarioTest
+
+
+class RoleScenarioTest(ScenarioTest):
+
+    def enable_large_payload(self, size=8192):
+        from azure_devtools.scenario_tests import LargeResponseBodyProcessor
```
Comment (tjprescott) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/tests/role_scenario_test.py]: We should consider moving this into the test base. It is used in many places to work around what I would consider a bug in devtools.

Comment (yugangw-msft) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/tests/role_scenario_test.py]: we can port https://github.com/Azure/azure-cli/issues/4127 over

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 6 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: The problem is GUID can be used as assignee names, so w/o graph query, we would not know. For people login with service principal, they need to have the option to bypass the graph query.
I have proposed to the fallback to object id as long as it is a guid, but it is possible we could assign by mistake when service principal 2's object id happens to be the service principal 1's name

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 4 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Let me log an info. It is an expected exception when login as a SP w/p graph permission, so warning might be too much.

**Pull request merged**